File: samples/snippets/nodejs-snippets/test/sample_tests.ts
// Copyright 2024 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { describe, before, after, test, beforeEach } from 'node:test';
import assert from 'node:assert';
import {
    TestContainer,
    StartedTestContainer,
    GenericContainer,
    PullPolicy
} from "testcontainers";
import createTables from "../src/create_tables";
import createConnection from "../src/create_connection";
import writeDataWithDml from "../src/write_data_with_dml";
import writeDataWithDmlBatch from "../src/write_data_with_dml_batch";
import writeDataWithCopy from "../src/write_data_with_copy";
import queryData from "../src/query_data";
import queryWithParameter from "../src/query_data_with_parameter";
import addColumn from "../src/add_column";
import ddlBatch from "../src/ddl_batch";
import updateDataWithCopy from "../src/update_data_with_copy";
import queryDataWithNewColumn from "../src/query_data_with_new_column";
import writeWithTransactionUsingDml from "../src/update_data_with_transaction";
import tags from "../src/tags";
import readOnlyTransaction from "../src/read_only_transaction";
import dataBoost from "../src/data_boost";
import partitionedDml from "../src/partitioned_dml";

const container: TestContainer = new GenericContainer("gcr.io/cloud-spanner-pg-adapter/pgadapter-emulator")
    .withExposedPorts(5432)
    .withPullPolicy(PullPolicy.alwaysPull());

describe('running samples', () => {
    let startedTestContainer: StartedTestContainer;
    const log = console.log;
    const loggedLines: string[] = [];

    before(async () => {
        startedTestContainer = await container.start();
        console.log = (...args: any[]) => {
            loggedLines.push(args.map(arg => typeof arg === 'string' ? arg : JSON.stringify(arg)).join(' '));
        };
    });

    after(async () => {
        console.log = log;
        if (startedTestContainer) {
            await startedTestContainer.stop({remove: true});
        }
    });

    beforeEach(() => {
        loggedLines.length = 0;
    });

    test('create tables', { timeout: 30000 }, async () => {
        await createTables(startedTestContainer.getHost(), startedTestContainer.getMappedPort(5432), "example-db");
        assert(loggedLines.includes("Created Singers & Albums tables in database: [example-db]"));
    });
    test('create connection', { timeout: 30000 }, async () => {
        await createConnection(startedTestContainer.getHost(), startedTestContainer.getMappedPort(5432), "example-db");
        assert(loggedLines.includes("Greeting from Cloud Spanner PostgreSQL: Hello world!"));
    });
    test('write data with DML', { timeout: 30000 }, async () => {
        await writeDataWithDml(startedTestContainer.getHost(), startedTestContainer.getMappedPort(5432), "example-db");
        assert(loggedLines.includes("4 records inserted"));
    });
    test('execute DML batch', { timeout: 30000 }, async () => {
        await writeDataWithDmlBatch(startedTestContainer.getHost(), startedTestContainer.getMappedPort(5432), "example-db");
        assert(loggedLines.includes("3 records inserted"));
    });
    test('copy from stdin', { timeout: 30000 }, async () => {
        await writeDataWithCopy(startedTestContainer.getHost(), startedTestContainer.getMappedPort(5432), "example-db");
        assert(loggedLines.includes("Copied 5 singers"));
        assert(loggedLines.includes("Copied 5 albums"));
    });
    test('query data', { timeout: 30000 }, async () => {
        await queryData(startedTestContainer.getHost(), startedTestContainer.getMappedPort(5432), "example-db");
        assert(loggedLines.includes("1 2 Go, Go, Go"));
        assert(loggedLines.includes("2 2 Forever Hold Your Peace"));
        assert(loggedLines.includes("1 1 Total Junk"));
        assert(loggedLines.includes("2 1 Green"));
        assert(loggedLines.includes("2 3 Terrified"));
    });
    test('query with parameter', { timeout: 30000 }, async () => {
        await queryWithParameter(startedTestContainer.getHost(), startedTestContainer.getMappedPort(5432), "example-db");
        assert(loggedLines.includes("12 Melissa Garcia"));
    });
    test('add column', { timeout: 30000 }, async () => {
        await addColumn(startedTestContainer.getHost(), startedTestContainer.getMappedPort(5432), "example-db");
        assert(loggedLines.includes("Added marketing_budget column"));
    });
    test('ddl batch', { timeout: 30000 }, async () => {
        await ddlBatch(startedTestContainer.getHost(), startedTestContainer.getMappedPort(5432), "example-db");
        assert(loggedLines.includes("Added venues and concerts tables"));
    });
    test('update data', { timeout: 30000 }, async () => {
        await updateDataWithCopy(startedTestContainer.getHost(), startedTestContainer.getMappedPort(5432), "example-db");
        assert(loggedLines.includes("Updated 2 albums"));
    });
    test('query data with new column', { timeout: 30000 }, async () => {
        await queryDataWithNewColumn(startedTestContainer.getHost(), startedTestContainer.getMappedPort(5432), "example-db");
        assert(loggedLines.includes("1 1 100000"));
        assert(loggedLines.includes("1 2 null"));
        assert(loggedLines.includes("2 1 null"));
        assert(loggedLines.includes("2 2 500000"));
        assert(loggedLines.includes("2 3 null"));
    });
    test('update data with transaction', { timeout: 30000 }, async () => {
        await writeWithTransactionUsingDml(startedTestContainer.getHost(), startedTestContainer.getMappedPort(5432), "example-db");
        assert(loggedLines.includes("Transferred marketing budget from Album 2 to Album 1"));
    });
    test('transaction and statement tags', { timeout: 30000 }, async () => {
        await tags(startedTestContainer.getHost(), startedTestContainer.getMappedPort(5432), "example-db");
        assert(loggedLines.includes("Reduced marketing budget"));
    });
    test('read-only transaction', { timeout: 30000 }, async () => {
        await readOnlyTransaction(startedTestContainer.getHost(), startedTestContainer.getMappedPort(5432), "example-db");
        assert(loggedLines.includes("1 1 Total Junk"));
        assert(loggedLines.includes("1 2 Go, Go, Go"));
        assert(loggedLines.includes("2 1 Green"));
        assert(loggedLines.includes("2 2 Forever Hold Your Peace"));
        assert(loggedLines.includes("2 3 Terrified"));
    });
    test('data boost', { timeout: 30000 }, async () => {
        await dataBoost(startedTestContainer.getHost(), startedTestContainer.getMappedPort(5432), "example-db");
        assert(loggedLines.includes("2 Catalina Smith"));
        assert(loggedLines.includes("4 Lea Martin"));
        assert(loggedLines.includes("12 Melissa Garcia"));
        assert(loggedLines.includes("14 Jacqueline Long"));
        assert(loggedLines.includes("16 Sarah Wilson"));
        assert(loggedLines.includes("18 Maya Patel"));
        assert(loggedLines.includes("1 Marc Richards"));
        assert(loggedLines.includes("3 Alice Trentor"));
        assert(loggedLines.includes("5 David Lomond"));
        assert(loggedLines.includes("13 Russel Morales"));
        assert(loggedLines.includes("15 Dylan Shaw"));
        assert(loggedLines.includes("17 Ethan Miller"));
    });
    test('partitioned DML', { timeout: 30000 }, async () => {
        await partitionedDml(startedTestContainer.getHost(), startedTestContainer.getMappedPort(5432), "example-db");
        assert(loggedLines.includes("Updated at least 3 albums"));
    });
});

File: samples/snippets/dotnet-snippets/sample-tests/SampleTests.cs
// Copyright 2024 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using dotnet_snippets;
using DotNet.Testcontainers.Builders;
using DotNet.Testcontainers.Containers;
using DotNet.Testcontainers.Images;

namespace sample_tests;

[TestFixture]
public class SampleTests
{
    private const string Database = "example-db";
    private IContainer _container;
    
    [OneTimeSetUp]
    public async Task StartEmulator()
    {
        _container = await StartPGAdapterAndEmulator();
    }

    [OneTimeTearDown]
    public async Task StopEmulator()
    {
        await _container.StopAsync();
    }

    string CaptureOutput(Action action)
    {
        var original = Console.Out;
        try
        {
            using var sw = new StringWriter();
            Console.SetOut(sw);
            action.Invoke();
            return sw.ToString();
        }
        finally
        {
            Console.SetOut(original);
        }
    }
    
    [Test, Order(1)]
    public void CreateTables()
    {
        var output = CaptureOutput(() =>
        CreateTablesSample.CreateTables(_container.Hostname, _container.GetMappedPublicPort(5432), Database));
        var expected = $"Created Singers & Albums tables in database: [{Database}]{Environment.NewLine}";
        Assert.That(output, Is.EqualTo(expected));
    }
    
    [Test, Order(2)]
    public void CreateConnection()
    {
        var output = CaptureOutput(() =>
            CreateConnectionSample.CreateConnection(_container.Hostname, _container.GetMappedPublicPort(5432), Database));
        var expected = $"Greeting from Cloud Spanner PostgreSQL: Hello World!{Environment.NewLine}";
        Assert.That(output, Is.EqualTo(expected));
    }
    
    [Test, Order(3)]
    public void WriteDataWithDml()
    {
        var output = CaptureOutput(() =>
            WriteDataWithDmlSample.WriteDataWithDml(_container.Hostname, _container.GetMappedPublicPort(5432), Database));
        var expected = $"4 records inserted.{Environment.NewLine}";
        Assert.That(output, Is.EqualTo(expected));
    }
    
    [Test, Order(4)]
    public void WriteDataWithDmlBatch()
    {
        var output = CaptureOutput(() =>
            WriteDataWithDmlBatchSample.WriteDataWithDmlBatch(_container.Hostname, _container.GetMappedPublicPort(5432), Database));
        var expected = $"3 records inserted.{Environment.NewLine}";
        Assert.That(output, Is.EqualTo(expected));
    }
    
    [Test, Order(5)]
    public void WriteDataWithCopy()
    {
        var output = CaptureOutput(() =>
            WriteDataWithCopySample.WriteDataWithCopy(_container.Hostname, _container.GetMappedPublicPort(5432), Database));
        var expected = $"Copied 5 singers{Environment.NewLine}Copied 5 albums{Environment.NewLine}";
        Assert.That(output, Is.EqualTo(expected));
    }
    
    [Test, Order(6)]
    public void QueryData()
    {
        var output = CaptureOutput(() =>
            QueryDataSample.QueryData(_container.Hostname, _container.GetMappedPublicPort(5432), Database));
        var expected = $"1 2 Go, Go, Go{Environment.NewLine}" +
                       $"2 2 Forever Hold Your Peace{Environment.NewLine}" +
                       $"1 1 Total Junk{Environment.NewLine}" +
                       $"2 1 Green{Environment.NewLine}" +
                       $"2 3 Terrified{Environment.NewLine}";
        Assert.That(output, Is.EqualTo(expected));
    }
    
    [Test, Order(7)]
    public void QueryDataWithParameter()
    {
        var output = CaptureOutput(() =>
            QueryDataWithParameterSample.QueryDataWithParameter(_container.Hostname, _container.GetMappedPublicPort(5432), Database));
        var expected = $"12 Melissa Garcia{Environment.NewLine}";
        Assert.That(output, Is.EqualTo(expected));
    }
    
    [Test, Order(8)]
    public void QueryWithTimeout()
    {
        var output = CaptureOutput(() =>
            QueryWithTimeoutSample.QueryWithTimeout(_container.Hostname, _container.GetMappedPublicPort(5432), Database));
        Assert.That(output, Is.EqualTo(""));
    }
    
    [Test, Order(9)]
    public void AddColumn()
    {
        var output = CaptureOutput(() =>
            AddColumnSample.AddColumn(_container.Hostname, _container.GetMappedPublicPort(5432), Database));
        Assert.That(output, Is.EqualTo($"Added marketing_budget column{Environment.NewLine}"));
    }
    
    [Test, Order(10)]
    public void DdlBatch()
    {
        var output = CaptureOutput(() =>
            DdlBatchSample.DdlBatch(_container.Hostname, _container.GetMappedPublicPort(5432), Database));
        Assert.That(output, Is.EqualTo($"Added venues and concerts tables{Environment.NewLine}"));
    }
    
    [Test, Order(11)]
    public void UpdateDataWithCopy()
    {
        var output = CaptureOutput(() =>
            UpdateDataWithCopySample.UpdateDataWithCopy(_container.Hostname, _container.GetMappedPublicPort(5432), Database));
        Assert.That(output, Is.EqualTo($"Updated 2 albums{Environment.NewLine}"));
    }
    
    [Test, Order(12)]
    public void QueryDataWithNewColumn()
    {
        var output = CaptureOutput(() =>
            QueryDataWithNewColumnSample.QueryWithNewColumnData(_container.Hostname, _container.GetMappedPublicPort(5432), Database));
        Assert.That(output, Is.EqualTo($"1 1 100000{Environment.NewLine}" +
                                       $"1 2 {Environment.NewLine}" +
                                       $"2 1 {Environment.NewLine}" +
                                       $"2 2 500000{Environment.NewLine}" +
                                       $"2 3 {Environment.NewLine}"));
    }
    
    [Test, Order(13)]
    public void WriteDataWithTransaction()
    {
        var output = CaptureOutput(() =>
            UpdateDataWithTransactionSample.WriteWithTransactionUsingDml(_container.Hostname, _container.GetMappedPublicPort(5432), Database));
        Assert.That(output, Is.EqualTo($"Transferred marketing budget from Album 2 to Album 1{Environment.NewLine}"));
    }
    
    [Test, Order(14)]
    public void Tags()
    {
        var output = CaptureOutput(() =>
            TagsSample.Tags(_container.Hostname, _container.GetMappedPublicPort(5432), Database));
        Assert.That(output, Is.EqualTo($"Reduced marketing budget{Environment.NewLine}"));
    }
    
    [Test, Order(15)]
    public void ReadOnlyTransaction()
    {
        var output = CaptureOutput(() =>
            ReadOnlyTransactionSample.ReadOnlyTransaction(_container.Hostname, _container.GetMappedPublicPort(5432), Database));
        Assert.That(output, Is.EqualTo($"1 1 Total Junk{Environment.NewLine}"
                                       + $"1 2 Go, Go, Go{Environment.NewLine}"
                                       + $"2 1 Green{Environment.NewLine}"
                                       + $"2 2 Forever Hold Your Peace{Environment.NewLine}"
                                       + $"2 3 Terrified{Environment.NewLine}"
                                       + $"2 2 Forever Hold Your Peace{Environment.NewLine}"
                                       + $"1 2 Go, Go, Go{Environment.NewLine}"
                                       + $"2 1 Green{Environment.NewLine}"
                                       + $"2 3 Terrified{Environment.NewLine}"
                                       + $"1 1 Total Junk{Environment.NewLine}"));
    }
    
    [Test, Order(16)]
    public void DataBoost()
    {
        var output = CaptureOutput(() =>
            DataBoostSample.DataBoost(_container.Hostname, _container.GetMappedPublicPort(5432), Database));
        Assert.That(output, Is.EqualTo($"2 Catalina Smith{Environment.NewLine}"
                                       + $"4 Lea Martin{Environment.NewLine}"
                                       + $"12 Melissa Garcia{Environment.NewLine}"
                                       + $"14 Jacqueline Long{Environment.NewLine}"
                                       + $"16 Sarah Wilson{Environment.NewLine}"
                                       + $"18 Maya Patel{Environment.NewLine}"
                                       + $"1 Marc Richards{Environment.NewLine}"
                                       + $"3 Alice Trentor{Environment.NewLine}"
                                       + $"5 David Lomond{Environment.NewLine}"
                                       + $"13 Russel Morales{Environment.NewLine}"
                                       + $"15 Dylan Shaw{Environment.NewLine}"
                                       + $"17 Ethan Miller{Environment.NewLine}"));
    }
        
    [Test, Order(17)]
    public void PartitionedDml()
    {
        var output = CaptureOutput(() =>
            PartitionedDmlSample.PartitionedDml(_container.Hostname, _container.GetMappedPublicPort(5432), Database));
        Assert.That(output, Is.EqualTo($"Updated at least 3 albums{Environment.NewLine}"));
    }

    private static async Task<IContainer> StartPGAdapterAndEmulator()
    {
        var container = new ContainerBuilder()
            .WithImagePullPolicy(imagePullPolicy: PullPolicy.Always)
            // Use the PGAdapter image that also contains automatically starts the Spanner emulator.
            .WithImage("gcr.io/cloud-spanner-pg-adapter/pgadapter-emulator")
            // Bind port 5432 of the container to a random port on the host.
            .WithPortBinding(5432, true)
            // Wait until the PostgreSQL port is available.
            .WithWaitStrategy(Wait.ForUnixContainer().UntilPortIsAvailable(5432))
            // Build the container configuration.
            .Build();
        await container.StartAsync();
        return container;
    }
    
}